### ISSUE TITLE
FAQ? Why doesn't this use dmsetup/cryptsetup ?

### ISSUE THREAD

**Issue by rdebath** [Aorimn/dislocker#32]
The filesystem image file within a fuse mount is a rather unusual way of implementing what amounts to a block device 'shim'.  IMO a network block device simulator (like qemu-nbd) would be a more usual way of doing exactly this. However, using the kernel encryption routines in the way that truecrypt on Linux does would be the most 'correct' (and fastest) solution, if it's possible. (Truecrypt just sets up a dmsetup/cryptsetup mapping for the device then steps aside and lets the kernel do all the heavy lifting).

Obviously, Microsoft have an almost terminal case of "not invented here" which will probably prevent the kernel solution without adding special purpose code to the kernel itself.

But, I would like to see something in the repository to document why these choices were made, and perhaps any you may foresee changing. (eg: new kernel module)

So in short, how about starting a FAQ ?

**Comment by Aorimn** [comment]
FUSE had multiple advantages at the time dislocker was written:
* I knew its existence - whereas network block devices I didn't, for example, or at least not in this context
* It was straight forward to use, it wouldn't be the problem in the implementation
* It was in userspace, so easier to test and wouldn't crash the kernel if faulting in some way
* Plus, the NTFS part was using FUSE, so it seemed logical.

Now, I've done some work to extract the core code into a library. The current dislocker implementation is a quite small program which actually uses this library. Therefore, I don't think making a network block device would be too difficult to implement.
It also makes sense to use dmcrypt/cryptsetup to implement it, BitLocker really has some resemblance in LUKS itself. It might also gain in performance. The problem would be portability however.

In essence, I don't really have a plan on this road of changing from FUSE. However, I'm open to any suggestion - like the one you cited.

As for a FAQ, I would say "why not!" I'll try to make some entries in it.

**Comment by claunia** [comment]
Library pros:
.- Portability
.- Ability to embed in other software (data recovery e.g.)
.- No kernel panic

For optimization, that's neligible, imho, most of the performance benefits would come of optimizing the encryption code to use CPU extensions (ARM secure element, VIA Padlock, Intel AES-NI), and that can be done in user-space.

This is my suggestion.
